pkg/controllers/clusterinventory/clusterprofile/controller.go

```diff
@@ -0,0 +1,265 @@
+/*
+Copyright (c) Microsoft Corporation.
+Licensed under the MIT license.
+*/
+
+// Package clusterprofile features a controller to generate clusterprofile objects from MemberCluster
+package clusterprofile
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+	clusterinventory "sigs.k8s.io/cluster-inventory-api/apis/v1alpha1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	clusterv1beta1 "go.goms.io/fleet/apis/cluster/v1beta1"
+	"go.goms.io/fleet/pkg/utils"
+)
+
+const (
+	// clusterProfileCleanupFinalizer is the finalizer added to a MemberCluster object if
+	// a corresponding ClusterProfile object has been created.
+	clusterProfileCleanupFinalizer = "kubernetes-fleet.io/cluster-profile-cleanup"
+
+	// the list of reasons in the cluster profile status
+	clusterNoStatusReason      = "MemberAgentReportedNoStatus"
+	clusterHeartbeatLostReason = "MemberAgentHeartbeatLost"
+	clusterHealthUnknownReason = "MemberAgentReportedNoHealthInfo"
+	clusterUnHealthyReason     = "MemberClusterAPIServerUnhealthy"
+	clusterHealthyReason       = "MemberClusterAPIServerHealthy"
+)
+
+// Reconciler reconciles a MemberCluster object and creates the corresponding ClusterProfile
+// object in the designated namespace.
+type Reconciler struct {
+	client.Client
+	ClusterProfileNamespace   string
+	ClusterUnhealthyThreshold time.Duration
```

Comment:
can we create a structure like healthyCheckConfig?

we will have ControlPlaneUnhealthyThreshold there. Later we can add more.

Reply:
can we add when we have more parameter? I am not sure if there are going to be more?